// @sentry/capacitor/dist/build/transports/native.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeTransport = exports.DEFAULT_BUFFER_SIZE = void 0;
exports.makeNativeTransport = makeNativeTransport;
exports.makeNativeTransportFactory = makeNativeTransportFactory;
const core_1 = require("@sentry/core");
const wrapper_1 = require("../wrapper");
exports.DEFAULT_BUFFER_SIZE = 30;
/** Native Transport class implementation */
class NativeTransport {
    constructor(options = {}) {
        this._buffer = (0, core_1.makePromiseBuffer)(options.bufferSize || exports.DEFAULT_BUFFER_SIZE);
    }
    /**
     * Sends the envelope to the Store endpoint in Sentry.
     *
     * @param envelope Envelope that should be sent to Sentry.
     */
    send(envelope) {
        return this._buffer.add(() => wrapper_1.NATIVE.sendEnvelope(envelope));
    }
    /**
     * Wait for all envelopes to be sent or the timeout to expire, whichever comes first.
     *
     * @param timeout Maximum time in ms the transport should wait for envelopes to be flushed. Omitting this parameter will
     *   cause the transport to wait until all events are sent before resolving the promise.
     * @returns A promise that will resolve with `true` if all events are sent before the timeout, or `false` if there are
     * still events in the queue when the timeout is reached.
     */
    flush(timeout) {
        return this._buffer.drain(timeout);
    }
}
exports.NativeTransport = NativeTransport;
/**
 * Creates a Native Transport.
 */
function makeNativeTransport(options = {}) {
    return new NativeTransport(options);
}
/**
 * Creates a Native Transport factory if the native transport is available.
 */
function makeNativeTransportFactory({ enableNative, }) {
    if (enableNative && wrapper_1.NATIVE.isNativeClientAvailable()) {
        return makeNativeTransport;
    }
    return null;
}
