// @sentry/capacitor/dist/build/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = init;
exports.close = close;
exports.nativeCrash = nativeCrash;
const browser_1 = require("@sentry/browser");
const core_1 = require("@sentry/core");
const client_1 = require("./client");
const default_1 = require("./integrations/default");
const NativeLogListener_1 = require("./NativeLogListener");
const scopeSync_1 = require("./scopeSync");
const encodePolyfill_1 = require("./transports/encodePolyfill");
const native_1 = require("./transports/native");
const safeFactory_1 = require("./utils/safeFactory");
const textEncoder_1 = require("./utils/textEncoder");
const wrapper_1 = require("./wrapper");
/**
 * Initializes the Capacitor SDK alongside a sibling Sentry SDK
 * @param options Options for the SDK
 * @param originalInit The init function of the sibling SDK, leave blank to initialize with `@sentry/browser`
 */
function init(passedOptions, originalInit = browser_1.init) {
    var _a, _b, _c, _d, _e, _f, _g;
    /**
     * Shared options are the options that are shared between the browser and native SDKs.
     */
    const sharedOptions = {
        enableAutoSessionTracking: true,
        enableWatchdogTerminationTracking: true,
        enableCaptureFailedRequests: false,
        ...passedOptions,
    };
    sharedOptions.siblingOptions && delete sharedOptions.siblingOptions;
    if (sharedOptions.enabled === false || wrapper_1.NATIVE.platform === 'web') {
        sharedOptions.enableNative = false;
        sharedOptions.enableNativeNagger = false;
    }
    else {
        // keep the original value if user defined it.
        (_a = sharedOptions.enableNativeNagger) !== null && _a !== void 0 ? _a : (sharedOptions.enableNativeNagger = true);
        (_b = sharedOptions.enableNative) !== null && _b !== void 0 ? _b : (sharedOptions.enableNative = true);
    }
    if (sharedOptions.enableNative &&
        !passedOptions.transport &&
        wrapper_1.NATIVE.platform !== 'web') {
        sharedOptions.transport = passedOptions.transport || native_1.makeNativeTransport;
        sharedOptions.transportOptions = {
            ...((_c = passedOptions.transportOptions) !== null && _c !== void 0 ? _c : {}),
            bufferSize: native_1.DEFAULT_BUFFER_SIZE,
        };
    }
    if (!(0, textEncoder_1.IsTextEncoderAvailable)()) {
        (0, encodePolyfill_1.useEncodePolyfill)();
    }
    if (sharedOptions.enableNative) {
        (0, scopeSync_1.enableSyncToNative)((0, core_1.getGlobalScope)());
        (0, scopeSync_1.enableSyncToNative)((0, core_1.getIsolationScope)());
    }
    /**
     * Browser options are the options that are only used by the browser SDK.
     */
    const browserOptions = {
        ...(_d = passedOptions.siblingOptions) === null || _d === void 0 ? void 0 : _d.vueOptions,
        ...(_e = passedOptions.siblingOptions) === null || _e === void 0 ? void 0 : _e.nuxtClientOptions,
        ...sharedOptions,
        integrations: (0, safeFactory_1.safeFactory)(passedOptions.integrations, {
            loggerMessage: 'The integrations threw an error',
        }),
        enableMetrics: (_f = sharedOptions._experiments) === null || _f === void 0 ? void 0 : _f.enableMetrics,
        beforeSendMetric: (_g = sharedOptions._experiments) === null || _g === void 0 ? void 0 : _g.beforeSendMetric,
    };
    browserOptions.defaultIntegrations =
        passedOptions.defaultIntegrations === undefined
            ? (0, default_1.getDefaultIntegrations)(sharedOptions)
            : passedOptions.defaultIntegrations;
    /**
     * Mobile options are the options that are only used by the native SDK.
     */
    const mobileOptions = {
        ...sharedOptions,
        enableAutoSessionTracking: wrapper_1.NATIVE.platform !== 'web' && sharedOptions.enableAutoSessionTracking,
    };
    (0, client_1.sdkInit)(browserOptions, mobileOptions, originalInit, passedOptions.transport);
}
/**
 * Closes the SDK, stops sending events.
 */
async function close() {
    (0, NativeLogListener_1.stopNativeLogListener)();
    try {
        const client = (0, core_1.getClient)();
        if (client) {
            await client.close();
            await wrapper_1.NATIVE.closeNativeSdk();
        }
    }
    catch (e) {
        core_1.debug.error('Failed to close the SDK');
    }
}
/**
 * If native client is available it will trigger a native crash
 * Use this only for testing purposes
 */
function nativeCrash() {
    wrapper_1.NATIVE.crash();
}
