// @sentry/capacitor/dist/build/wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NATIVE = void 0;
exports.pauseAppHangTracking = pauseAppHangTracking;
exports.resumeAppHangTracking = resumeAppHangTracking;
/* eslint-disable max-lines */
const core_1 = require("@capacitor/core");
const core_2 = require("@sentry/core");
const nativeOptions_1 = require("./nativeOptions");
const plugin_1 = require("./plugin");
const normalize_1 = require("./utils/normalize");
const vendor_1 = require("./vendor");
const fromByteArray_1 = require("./vendor/fromByteArray");
/**
 * Internal interface for calling native functions
 */
exports.NATIVE = {
    /**
     * Sending the event over the bridge to native
     * @param event Event
     */
    async sendEnvelope(envelope) {
        if (!this.enableNative) {
            throw this._DisabledNativeError;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        const EOL = 10; // utf8ToBytes('\n');
        const [envelopeHeader, envelopeItems] = envelope;
        const headerString = JSON.stringify(envelopeHeader);
        let envelopeBytes = (0, vendor_1.utf8ToBytes)(headerString);
        envelopeBytes.push(EOL);
        for (const rawItem of envelopeItems) {
            const [itemHeader, itemPayload] = this._processItem(rawItem);
            let bytesContentType;
            let bytesPayload = [];
            if (typeof itemPayload === 'string') {
                bytesContentType = 'text/plain';
                bytesPayload = (0, vendor_1.utf8ToBytes)(itemPayload);
            }
            else if (itemPayload instanceof Uint8Array) {
                bytesContentType =
                    typeof itemHeader.content_type === 'string'
                        ? itemHeader.content_type
                        : 'application/octet-stream';
                bytesPayload = [...itemPayload];
            }
            else {
                bytesContentType = typeof itemHeader.content_type === 'string' ? itemHeader.content_type : 'application/json';
                bytesPayload = (0, vendor_1.utf8ToBytes)(JSON.stringify(itemPayload));
            }
            // Content type is not inside BaseEnvelopeItemHeaders.
            itemHeader.content_type = bytesContentType;
            itemHeader.length = bytesPayload.length;
            const serializedItemHeader = JSON.stringify(itemHeader);
            envelopeBytes.push(...(0, vendor_1.utf8ToBytes)(serializedItemHeader));
            envelopeBytes.push(EOL);
            envelopeBytes = envelopeBytes.concat(bytesPayload);
            envelopeBytes.push(EOL);
        }
        let transportStatusCode = 200;
        await plugin_1.SentryCapacitor.captureEnvelope({ envelope: (0, fromByteArray_1.base64StringFromByteArray)(envelopeBytes) }).then(_ => _, // We only want to know if it failed.
        // We only want to know if it failed.
        failed => {
            core_2.debug.error('Failed to capture Envelope: ', failed);
            transportStatusCode = 500;
        });
        return { statusCode: transportStatusCode };
    },
    /**
     * Starts native with the provided options
     * @param options CapacitorOptions
     */
    async initNativeSdk(options) {
        if (!options.dsn) {
            core_2.debug.warn('Warning: No DSN was provided. The Sentry SDK will be disabled. Native SDK will also not be initalized.');
            this.enableNative = false;
            return false;
        }
        if (!options.enableNative) {
            if (options.enableNativeNagger) {
                core_2.debug.warn('Note: Native Sentry SDK is disabled.');
            }
            this.enableNative = false;
            return false;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        // filter out all options that would crash native
        const filteredOptions = (0, nativeOptions_1.FilterNativeOptions)(options);
        return plugin_1.SentryCapacitor.initNativeSdk({ options: filteredOptions });
    },
    /**
     * Fetches the device contexts. Not used on Android.
     */
    async fetchNativeDeviceContexts() {
        if (!this.enableNative) {
            throw this._DisabledNativeError;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        return plugin_1.SentryCapacitor.fetchNativeDeviceContexts();
    },
    /**
     * Fetches native log attributes for log enrichment.
     */
    async fetchNativeLogAttributes() {
        if (!this.enableNative) {
            throw this._DisabledNativeError;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        return plugin_1.SentryCapacitor.fetchNativeLogAttributes();
    },
    /**
     * Fetches the release from native
     */
    async fetchNativeRelease() {
        if (!this.enableNative) {
            throw this._DisabledNativeError;
        }
        return plugin_1.SentryCapacitor.fetchNativeRelease();
    },
    async fetchNativeSdkInfo() {
        if (!this.enableNative) {
            throw this._DisabledNativeError;
        }
        return plugin_1.SentryCapacitor.fetchNativeSdkInfo();
    },
    /**
     * Triggers a native crash.
     * Use this only for testing purposes.
     */
    crash() {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        plugin_1.SentryCapacitor.crash();
    },
    /**
     * Sets the user in the native scope.
     * Passing null clears the user.
     * @param key string
     * @param value string
     */
    setUser(user) {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        // separate and serialze all non-default user keys.
        let defaultUserKeys = null;
        let otherUserKeys = null;
        if (user) {
            const { id, ip_address, email, username, ...otherKeys } = user;
            defaultUserKeys =
                this._serializeObject({
                    email,
                    id,
                    ip_address,
                    username,
                }, true);
            otherUserKeys = this._serializeObject(otherKeys, true);
        }
        plugin_1.SentryCapacitor.setUser({ defaultUserKeys, otherUserKeys });
    },
    /**
     * Sets a tag in the native module.
     * @param key string
     * @param value string
     */
    setTag(key, value) {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        const stringifiedValue = typeof value === 'string' ? value : JSON.stringify(value);
        plugin_1.SentryCapacitor.setTag({ key, value: stringifiedValue });
    },
    /**
     * Sets an extra in the native scope, will stringify
     * extra value if it isn't already a string.
     * @param key string
     * @param extra any
     */
    setExtra(key, extra) {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        // we stringify the extra as native only takes in strings.
        const stringifiedValue = typeof extra === 'string' ? extra : JSON.stringify(extra);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        plugin_1.SentryCapacitor.setExtra({ key, value: stringifiedValue });
    },
    /**
     * Adds breadcrumb to the native scope.
     * @param breadcrumb Breadcrumb
     */
    addBreadcrumb(breadcrumb) {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        plugin_1.SentryCapacitor.addBreadcrumb({
            ...breadcrumb,
            // Process and convert deprecated levels
            level: breadcrumb.level
                ? this._processLevel(breadcrumb.level)
                : undefined,
            data: breadcrumb.data
                ? (0, normalize_1.convertToNormalizedObject)(breadcrumb.data)
                : undefined,
        });
    },
    /**
     * Clears breadcrumbs on the native scope.
     */
    clearBreadcrumbs() {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        plugin_1.SentryCapacitor.clearBreadcrumbs();
    },
    /**
     * Closes the Native Layer SDK
     */
    closeNativeSdk() {
        if (!this.enableNative) {
            // Only log and avoid rejecting.
            core_2.debug.log(this._DisabledNativeError);
            return Promise.resolve();
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        return plugin_1.SentryCapacitor.closeNativeSdk().then(() => {
            this.enableNative = false;
        });
    },
    /**
     * Sets context on the native scope. Not implemented in Android yet.
     * @param key string
     * @param context key-value map
     */
    setContext(key, context) {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        if (this.platform === 'android') {
            // setContext not available on the Android SDK yet.
            this.setExtra(key, context);
        }
        else {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            plugin_1.SentryCapacitor.setContext({
                key,
                value: context !== null ? this._serializeObject(context) : null,
            });
        }
    },
    pauseAppHangTracking() {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        else if (this.platform === 'ios') {
            plugin_1.SentryCapacitor.pauseAppHangTracking();
        }
        else {
            core_2.debug.warn(`PauseAppHangTracking ${this._IosOnlyMessage}`);
        }
    },
    resumeAppHangTracking() {
        if (!this.enableNative) {
            return;
        }
        if (!this.isNativeClientAvailable()) {
            throw this._NativeClientError;
        }
        else if (this.platform === 'ios') {
            plugin_1.SentryCapacitor.resumeAppHangTracking();
        }
        else {
            core_2.debug.warn(`ResumeAppHangTracking ${this._IosOnlyMessage}`);
        }
    },
    /**
     * Gets the event from envelopeItem and applies the level filter to the selected event.
     * @param data An envelope item containing the event.
     * @returns The event from envelopeItem or undefined.
     */
    _processItem(item) {
        if (exports.NATIVE.platform === 'android') {
            const [itemHeader, itemPayload] = item;
            if (itemHeader.type == 'event' || itemHeader.type == 'transaction') {
                const event = this._processLevels(itemPayload);
                if ('message' in event) {
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore Android still uses the old message object, without this the serialization of events will break.
                    event.message = { message: event.message };
                }
                return [itemHeader, event];
            }
        }
        return item;
    },
    /**
     * Serializes all values of root-level keys into strings.
     * @param data key-value map.
     * @returns An object where all root-level values are strings.
     */
    _serializeObject(data, removeUndefinedValues) {
        const serialized = {};
        Object.keys(data).forEach(dataKey => {
            const value = data[dataKey];
            if (removeUndefinedValues !== true || value !== undefined)
                serialized[dataKey] =
                    typeof value === 'string' ? value : JSON.stringify(value);
        });
        return serialized;
    },
    /**
     * Convert js severity level in event.level and event.breadcrumbs to more widely supported levels.
     * @param event
     * @returns Event with more widely supported Severity level strings
     */
    _processLevels(event) {
        var _a;
        const processed = {
            ...event,
            level: event.level ? this._processLevel(event.level) : undefined,
            breadcrumbs: (_a = event.breadcrumbs) === null || _a === void 0 ? void 0 : _a.map(breadcrumb => ({
                ...breadcrumb,
                level: breadcrumb.level
                    ? this._processLevel(breadcrumb.level)
                    : undefined,
            })),
        };
        return processed;
    },
    /**
     * Convert js severity level which has critical and log to more widely supported levels.
     * @param level
     * @returns More widely supported Severity level strings
     */
    _processLevel(level) {
        if (level == 'log') {
            return 'debug';
        }
        else if (level == 'critical') {
            return 'fatal';
        }
        return level;
    },
    /**
     * Checks whether the SentryCapacitor module is loaded.
     */
    isModuleLoaded() {
        return !!plugin_1.SentryCapacitor;
    },
    /**
     *  Checks whether the SentryCapacitor module is loaded and the native client is available
     */
    isNativeClientAvailable() {
        return (this.isModuleLoaded() && core_1.Capacitor.isPluginAvailable('SentryCapacitor'));
    },
    _DisabledNativeError: new Error('Native is disabled'),
    _NativeClientError: new Error("Native Client is not available, can't start on native."),
    _IosOnlyMessage: 'is only supported on iOS.',
    enableNative: true,
    platform: core_1.Capacitor.getPlatform(),
};
/**
 * When called, app hangs will be ignored until you call `ResumeAppHangTracking`.
 */
function pauseAppHangTracking() {
    exports.NATIVE.pauseAppHangTracking();
}
/**
 * Resumes sending detected app hangs to Sentry that were previsously ignored by `PauseAppHangTracking`.
 */
function resumeAppHangTracking() {
    exports.NATIVE.resumeAppHangTracking();
}
