// @sentry/capacitor/dist/esm/sdk.js
import { init as browserInit } from '@sentry/browser';
import { debug, getClient, getGlobalScope, getIsolationScope, } from '@sentry/core';
import { sdkInit } from './client';
import { getDefaultIntegrations } from './integrations/default';
import { stopNativeLogListener } from './NativeLogListener';
import { enableSyncToNative } from './scopeSync';
import { useEncodePolyfill } from './transports/encodePolyfill';
import { DEFAULT_BUFFER_SIZE, makeNativeTransport } from './transports/native';
import { safeFactory } from './utils/safeFactory';
import { IsTextEncoderAvailable } from './utils/textEncoder';
import { NATIVE } from './wrapper';
/**
 * Initializes the Capacitor SDK alongside a sibling Sentry SDK
 * @param options Options for the SDK
 * @param originalInit The init function of the sibling SDK, leave blank to initialize with `@sentry/browser`
 */
export function init(passedOptions, originalInit = browserInit) {
    var _a, _b, _c, _d, _e, _f, _g;
    /**
     * Shared options are the options that are shared between the browser and native SDKs.
     */
    const sharedOptions = {
        enableAutoSessionTracking: true,
        enableWatchdogTerminationTracking: true,
        enableCaptureFailedRequests: false,
        ...passedOptions,
    };
    sharedOptions.siblingOptions && delete sharedOptions.siblingOptions;
    if (sharedOptions.enabled === false || NATIVE.platform === 'web') {
        sharedOptions.enableNative = false;
        sharedOptions.enableNativeNagger = false;
    }
    else {
        // keep the original value if user defined it.
        (_a = sharedOptions.enableNativeNagger) !== null && _a !== void 0 ? _a : (sharedOptions.enableNativeNagger = true);
        (_b = sharedOptions.enableNative) !== null && _b !== void 0 ? _b : (sharedOptions.enableNative = true);
    }
    if (sharedOptions.enableNative &&
        !passedOptions.transport &&
        NATIVE.platform !== 'web') {
        sharedOptions.transport = passedOptions.transport || makeNativeTransport;
        sharedOptions.transportOptions = {
            ...((_c = passedOptions.transportOptions) !== null && _c !== void 0 ? _c : {}),
            bufferSize: DEFAULT_BUFFER_SIZE,
        };
    }
    if (!IsTextEncoderAvailable()) {
        useEncodePolyfill();
    }
    if (sharedOptions.enableNative) {
        enableSyncToNative(getGlobalScope());
        enableSyncToNative(getIsolationScope());
    }
    /**
     * Browser options are the options that are only used by the browser SDK.
     */
    const browserOptions = {
        ...(_d = passedOptions.siblingOptions) === null || _d === void 0 ? void 0 : _d.vueOptions,
        ...(_e = passedOptions.siblingOptions) === null || _e === void 0 ? void 0 : _e.nuxtClientOptions,
        ...sharedOptions,
        integrations: safeFactory(passedOptions.integrations, {
            loggerMessage: 'The integrations threw an error',
        }),
        enableMetrics: (_f = sharedOptions._experiments) === null || _f === void 0 ? void 0 : _f.enableMetrics,
        beforeSendMetric: (_g = sharedOptions._experiments) === null || _g === void 0 ? void 0 : _g.beforeSendMetric,
    };
    browserOptions.defaultIntegrations =
        passedOptions.defaultIntegrations === undefined
            ? getDefaultIntegrations(sharedOptions)
            : passedOptions.defaultIntegrations;
    /**
     * Mobile options are the options that are only used by the native SDK.
     */
    const mobileOptions = {
        ...sharedOptions,
        enableAutoSessionTracking: NATIVE.platform !== 'web' && sharedOptions.enableAutoSessionTracking,
    };
    sdkInit(browserOptions, mobileOptions, originalInit, passedOptions.transport);
}
/**
 * Closes the SDK, stops sending events.
 */
export async function close() {
    stopNativeLogListener();
    try {
        const client = getClient();
        if (client) {
            await client.close();
            await NATIVE.closeNativeSdk();
        }
    }
    catch (e) {
        debug.error('Failed to close the SDK');
    }
}
/**
 * If native client is available it will trigger a native crash
 * Use this only for testing purposes
 */
export function nativeCrash() {
    NATIVE.crash();
}
