// @sentry/capacitor/dist/build/integrations/logEnricherIntegration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logEnricherIntegration = void 0;
const core_1 = require("@sentry/core");
const wrapper_1 = require("../wrapper");
const INTEGRATION_NAME = 'LogEnricher';
const logEnricherIntegration = () => {
    return {
        name: INTEGRATION_NAME,
        setup(client) {
            // Cache log context immediately on setup
            cacheLogContext().then(() => {
                client.on('beforeCaptureLog', (log) => {
                    processLog(log);
                });
            }, reason => {
                core_1.debug.log(reason);
            });
        },
    };
};
exports.logEnricherIntegration = logEnricherIntegration;
let NativeCache = undefined;
/**
 * Sets a log attribute if the value exists and the attribute key is not already present.
 *
 * @param logAttributes - The log attributes object to modify.
 * @param key - The attribute key to set.
 * @param value - The value to set (only sets if truthy and key not present).
 * @param setEvenIfPresent - Whether to set the attribute if it is present. Defaults to true.
 */
function setLogAttribute(logAttributes, key, value, setEvenIfPresent = true) {
    if (value && (!logAttributes[key] || setEvenIfPresent)) {
        logAttributes[key] = value;
    }
}
async function cacheLogContext() {
    var _a, _b, _c, _d, _e;
    try {
        const response = await wrapper_1.NATIVE.fetchNativeLogAttributes();
        NativeCache = {
            ...(((_a = response === null || response === void 0 ? void 0 : response.contexts) === null || _a === void 0 ? void 0 : _a.device) && {
                brand: (_b = response.contexts.device) === null || _b === void 0 ? void 0 : _b.brand,
                model: (_c = response.contexts.device) === null || _c === void 0 ? void 0 : _c.model,
                family: (_d = response.contexts.device) === null || _d === void 0 ? void 0 : _d.family,
            }),
            ...(((_e = response === null || response === void 0 ? void 0 : response.contexts) === null || _e === void 0 ? void 0 : _e.os) && {
                os: response.contexts.os.name,
                version: response.contexts.os.version,
            }),
            ...((response === null || response === void 0 ? void 0 : response.release) && {
                release: response.release,
            }),
        };
    }
    catch (e) {
        return Promise.reject(`[LOGS]: Failed to prepare attributes from Native Layer: ${e}`);
    }
    return Promise.resolve();
}
function processLog(log) {
    var _a;
    if (NativeCache === undefined) {
        return;
    }
    // Save log.attributes to a new variable
    const logAttributes = (_a = log.attributes) !== null && _a !== void 0 ? _a : {};
    // Use setLogAttribute with the variable instead of direct assignment
    setLogAttribute(logAttributes, 'device.brand', NativeCache.brand);
    setLogAttribute(logAttributes, 'device.model', NativeCache.model);
    setLogAttribute(logAttributes, 'device.family', NativeCache.family);
    setLogAttribute(logAttributes, 'os.name', NativeCache.os);
    setLogAttribute(logAttributes, 'os.version', NativeCache.version);
    setLogAttribute(logAttributes, 'sentry.release', NativeCache.release);
    // Set log.attributes to the variable
    log.attributes = logAttributes;
}
